### PR TITLE
Feature/eip 7002 system call

### PR DESCRIPTION
**Pull request by rjnrohit**
Fixes Closes Resolves #

_Please choose one of the keywords above to refer to the issue this PR solves followed by the issue number (e.g. Fixes #000). If no issue number, remove the line. Also, remove everything marked optional that is not applicable. Remove this note after reading._

## Changes

- _List the changes_

#### What types of changes does your code introduce?

- [ ] Bugfix (a non-breaking change that fixes an issue)
- [ ] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a change that causes existing functionality not to work as expected)
- [ ] Optimization
- [ ] Refactoring
- [ ] Documentation update
- [ ] Build-related changes
- [ ] Other: _Description_

#### Requires testing

- [ ] Yes
- [ ] No

#### If yes, did you write tests?

- [ ] Yes
- [ ] No

#### Notes on testing

_Optional. Remove if not applicable._

#### Requires documentation update

- [ ] Yes
- [ ] No

_If yes, link the PR to the docs update or the issue with the details labeled `docs`. Remove if not applicable._

#### Requires explanation in Release Notes

- [ ] Yes
- [ ] No

_If yes, fill in the details here. Remove if not applicable._

## Remarks

_Optional. Remove if not applicable._

### CODE REVIEW

File: src/Nethermind/Nethermind.Consensus/Requests/DepositsProcessor.cs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+using Nethermind.Core.ConsensusRequests;
+using Nethermind.Core.Specs;
+using System.Linq;
+using Nethermind.Core.Extensions;
+
+namespace Nethermind.Consensus.Requests;
+
+public class DepositsProcessor : IDepositsProcessor
+{
+    private AbiSignature depositEventABI = new("DepositEvent", AbiType.DynamicBytes, AbiType.DynamicBytes, AbiType.DynamicBytes, AbiType.DynamicBytes, AbiType.DynamicBytes);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Consensus/Requests/DepositsProcessor.cs:16]: use Abi.UInt64 for relevant fields!

File: src/Nethermind/Nethermind.Consensus/Requests/DepositsProcessor.cs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    AbiEncoder abiEncoder = new();
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Consensus/Requests/DepositsProcessor.cs:17]: Can use AbiEncoder.Instance

File: src/Nethermind/Nethermind.Consensus/Requests/WithdrawalRequestsProcessor.cs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+        int sizeOfClass = 20 + 48 + 8;
+        int count = result.Length / sizeOfClass;
+        for (int i = 0; i < count; ++i)
+        {
+            WithdrawalRequest request = new();
+            Span<byte> span = new Span<byte>(result, i * sizeOfClass, sizeOfClass);
+            request.SourceAddress = new Address(span.Slice(0, 20).ToArray());
+            request.ValidatorPubkey = span.Slice(20, 48).ToArray();
+            request.Amount = BitConverter.ToUInt64(span.Slice(68, 8).ToArray().Reverse().ToArray()); // ToDo Optimize
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Consensus/Requests/WithdrawalRequestsProcessor.cs:61]: Either stackalloc space for it or you can create Uint256 with correct indianess and cast it to ulong.

File: src/Nethermind/Nethermind.Consensus/Requests/WithdrawalRequestsProcessor.cs [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+            yield break;
+
+        int sizeOfClass = 20 + 48 + 8;
+        int count = result.Length / sizeOfClass;
+        for (int i = 0; i < count; ++i)
+        {
+            WithdrawalRequest request = new();
+            Span<byte> span = new Span<byte>(result, i * sizeOfClass, sizeOfClass);
+            request.SourceAddress = new Address(span.Slice(0, 20).ToArray());
+            request.ValidatorPubkey = span.Slice(20, 48).ToArray();
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Consensus/Requests/WithdrawalRequestsProcessor.cs:60]: If we could keep it as ReadOnlyMmory<>, we could avoid that ToArray, not sure if worth it

File: src/Nethermind/Nethermind.Consensus/Requests/DepositsProcessor.cs [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+                        LogEntry log = logEntries[index];
+                        if (log.LoggersAddress == spec.DepositContractAddress)
+                        {
+                            var result = abiEncoder.Decode(AbiEncodingStyle.None, depositEventABI, log.Data);
+
+                            var newDeposit = new Deposit()
+                            {
+                                PubKey = (byte[])result[0],
+                                WithdrawalCredentials = (byte[])result[1],
+                                Amount = ((byte[])result[2]).Reverse().ToArray().ToULongFromBigEndianByteArrayWithoutLeadingZeros(), // ToDo not optimal - optimize
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Consensus/Requests/DepositsProcessor.cs:39]: Either use Abi.Uint64 or just parse it as new `Uint256(result[2], true/false)` and cast to ulong

File: src/Nethermind/Nethermind.Consensus/Requests/DepositsProcessor.cs (hunk 1 above)
Comment (rjnrohit) [reply src/Nethermind/Nethermind.Consensus/Requests/DepositsProcessor.cs:16]: Every field is of `bytes` type: https://etherscan.io/address/0x00000000219ab540356cbb839cbe05303d7705fa#code

File: src/Nethermind/Nethermind.Consensus/Requests/WithdrawalRequestsProcessor.cs (hunk 4 above)
Comment (rjnrohit) [reply src/Nethermind/Nethermind.Consensus/Requests/WithdrawalRequestsProcessor.cs:60]: we can't avoid `ToArray`

File: src/Nethermind/Nethermind.Consensus/Requests/WithdrawalRequestsProcessor.cs (hunk 3 above)
Comment (rjnrohit) [reply src/Nethermind/Nethermind.Consensus/Requests/WithdrawalRequestsProcessor.cs:61]: done, thanks!

**Pull request closed without merging**